File: 剑指offer/32 - I. 从上到下打印二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[int]:
        if not root: return []
        res, queue = [], collections.deque()
        queue.append(root)
        while queue:
            node = queue.popleft()
            res.append(node.val)
            if node.left: queue.append(node.left)
            if node.right: queue.append(node.right)
        return res
        

#时间复杂度 O(N)： N为二叉树的节点数量，即BFS需循环N次。
#空间复杂度 O(N)： 最差情况下，即当树为平衡二叉树时，最多有N/2个树节点同时在queue中，使用O(N)大小的额外空间。

https://leetcode-cn.com/problems/cong-shang-dao-xia-da-yin-er-cha-shu-lcof/solution/mian-shi-ti-32-i-cong-shang-dao-xia-da-yin-er-ch-4/

    

File: 剑指offer/C++_Version/32 - I. 从上到下打印二叉树.cpp
//方法1
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> levelOrder(TreeNode* root) 
    {
        vector<int> res;
        queue<TreeNode*> que;
        que.push(root);

        if(!root) return res;
        while(!que.empty())
        {
            TreeNode *node = que.front();
            que.pop();
            if(node->left) 
                que.push(node->left);
            if(node->right) 
                que.push(node->right);
            res.push_back(node->val);
        }
        return res;
    }
};
